### **Soal 1 Pandas Data Frame**

In [2]:
from sklearn.datasets import make_regression
# generate dataset
X, y = make_regression(n_samples=100, n_features=100, n_informative=10)

In [3]:
import pandas as pd
import numpy as np

In [19]:
columns = ["X"+str(i) for i in range(0,100)]

In [20]:
df = pd.DataFrame(data=X, columns=columns)
df['Y'] = y

In [21]:
df

X0        X1        X2        X3        X4        X5        X6  \
0  -0.150510  1.198468 -1.308840 -1.362130  2.536479  1.109925  1.355349   
1  -1.375057  1.269397  0.073905 -1.530055 -0.369479 -0.285104 -0.726432   
2  -0.516057 -0.620111  0.993264  0.361703  0.848958  1.737933 -1.551422   
3  -0.521364 -0.093254 -0.586911  0.376801  0.154522  0.085970 -1.079926   
4   1.099390 -0.365805 -2.095901 -1.217385  0.617228 -1.777702  2.023502   
..       ...       ...       ...       ...       ...       ...       ...   
95  0.492964 -1.348083 -0.005425  0.571508 -0.413566 -0.460787 -1.017452   
96 -0.877782  0.940185 -1.937979  0.339788 -0.271078 -0.698950 -1.400961   
97  0.280482 -0.403133 -1.091916 -0.680183 -0.640216  0.489133  0.045728   
98  0.258493 -1.259163 -0.817604  0.593928  0.695236  1.687215 -0.757293   
99  0.431789 -1.377232  0.599728 -1.356971 -1.030240 -0.410826 -0.563444   

          X7        X8        X9  ...       X91       X92       X93       X94  \
0  -0.844632 -0.191067  0.130068  ...  0.832074 -0.533873 -0.938971 -0.749984   
1  -0.150028  0.464435 -0.163439  ...  0.058416  0.570904  0.446075 -0.854271   
2  -1.039335  0.329179 -0.718489  ... -0.446344 -0.466629  1.425262 -0.936810   
3   0.634227  0.064144  0.530526  ... -1.024064  1.376883  0.595903  1.073199   
4   0.498432  0.489946 -2.004419  ... -2.540980 -1.154297 -1.937577 -0.065642   
..       ...       ...       ...  ...       ...       ...       ...       ...   
95  0.885152  0.723044  0.138794  ...  2.701621  1.089098  0.557197  1.168611   
96 -1.551190  0.089929  0.869573  ...  0.367351  1.534197  1.187577  0.560915   
97  0.653568 -0.140921 -0.504309  ...  0.033874  1.047484  0.349964 -0.111867   
98  0.361779 -2.019553 -0.191123  ...  1.528880  0.832715 -0.030700  0.649699   
99  0.564834  0.573373 -0.977662  ... -1.295158  0.594648  0.131885 -0.459122   

         X95       X96       X97       X98       X99           Y  
0  -0.713826 -0.112443  1.001908 -0.179890 -0.198196  330.992043  
1  -1.280076 -0.090614  1.991633  0.632675 -0.651177 -102.757232  
2   0.189945  1.359160  0.071292 -0.124621  0.009518   12.890767  
3   0.392549 -1.154733 -0.698879  0.562965  1.246105 -138.896044  
4   0.274896 -0.794667 -1.376999 -0.545403  1.208006  104.664745  
..       ...       ...       ...       ...       ...         ...  
95 -0.166564  1.756713 -0.354593  0.743465  0.273641  253.236857  
96 -1.069436  0.722873 -1.011457 -0.166549 -0.903486   90.673345  
97 -0.450053 -0.551052 -1.206955  0.016638 -0.964637  -21.910135  
98  1.486516  1.798980  1.879038 -1.208879 -0.357944  308.032303  
99  0.487189  2.044579  0.298438  1.736695 -0.950197  244.857417  

[100 rows x 101 columns]

Buatlah sebuah Dataframe dengan data di atas. Dimana terdapat variabel X dan y. buatlah menjadi column X0-X99 dan y pada akhir column.

In [2]:
#jangan di run
#cell contoh hasil

X0        X1        X2        X3        X4        X5        X6  \
0  -0.915212  1.493571  0.436189  0.705495 -1.761485 -1.249118 -0.407565   
1   0.878028 -0.140705 -0.944289 -0.174627  0.567625  0.300732  0.289273   
2   0.964140 -0.793125  0.841936  0.313088  0.199498  2.282574 -0.863335   
3  -0.882787 -1.152892  0.068783 -0.930178  0.036717 -0.197270  1.247525   
4   0.290320  0.316385 -2.246657  0.011012  0.311103  0.109057  0.308782   
..       ...       ...       ...       ...       ...       ...       ...   
95 -1.536589 -0.730357  0.779719  1.277046  1.321476 -1.119201 -0.092392   
96 -0.704000 -1.611241  1.814556  0.953647  0.076468  0.629419 -0.569326   
97  1.637832 -0.153730 -0.483074  1.892856  0.389054 -0.717983 -0.983700   
98 -1.167298 -0.260727  0.346849  1.693973 -0.097193  0.818206  0.813115   
99 -2.870851  0.489295 -0.686917  1.338321 -0.792317 -0.361618 -2.172565   

          X7        X8        X9  ...       X91       X92       X93       X94  \
0  -0.009842 -1.735053  0.910152  ... -0.621775  0.814272 -1.041435  1.895063   
1  -1.289092  0.686546  0.637312  ... -0.554019  0.193415  0.722733 -1.730579   
2  -0.191104  0.495262  0.223180  ... -0.262799 -2.593864  1.601533  1.455797   
3   0.474067 -0.704820 -1.593972  ...  0.092427  1.249872  2.186223 -0.131910   
4  -0.624079 -0.229622 -0.667855  ...  1.719986  2.142118  0.014183  1.037259   
..       ...       ...       ...  ...       ...       ...       ...       ...   
95  0.368560 -1.291459 -1.186540  ... -0.457575 -0.791506 -1.389717 -0.698408   
96  0.107478  1.417058 -1.063939  ... -2.022347 -0.064499  0.432952  0.398392   
97  0.753492  0.191647  0.197114  ...  0.241908  0.035325 -0.061961 -1.520946   
98 -0.566283 -1.620272 -0.536863  ...  0.324929  0.906779 -1.344463  0.633542   
99 -0.478471  0.227533 -1.606074  ...  0.425045 -1.007069 -0.782855  0.611999   

         X95       X96       X97       X98       X99           Y  
0  -1.436868  1.454175  0.394605  0.124696  1.926075 -100.608141  
1  -0.406840  2.340087 -0.114712  0.158274 -1.454759    1.689581  
2   0.058587  0.953225  0.752472  0.564658  1.131241  116.855979  
3   2.134867 -0.717014  0.083398 -0.772829  1.231793 -215.797817  
4  -1.227784 -1.471667 -0.911379 -0.068846  0.284262  197.679368  
..       ...       ...       ...       ...       ...         ...  
95 -0.989968  0.887914 -0.629646  0.654470 -1.453955 -109.582037  
96  1.301282  0.239558  1.705920  1.855406 -0.772279  177.805858  
97 -0.786646 -0.940910 -1.611571 -1.016891 -1.083254   36.836594  
98 -0.073089 -0.213599 -0.345081 -0.694501 -0.164250  390.464302  
99 -0.875159 -0.691232 -0.594701 -1.282145 -1.013643  133.545897  

[100 rows x 101 columns]

### Soal 02

Berapa banyak perenang yang memiliki medali perunggu(bronze)?

In [1]:
import pandas as pd
df=pd.read_csv('summer.csv')
df.head()

Year    City     Sport Discipline             Athlete Country Gender  \
0  1896  Athens  Aquatics   Swimming       HAJOS, Alfred     HUN    Men   
1  1896  Athens  Aquatics   Swimming    HERSCHMANN, Otto     AUT    Men   
2  1896  Athens  Aquatics   Swimming   DRIVAS, Dimitrios     GRE    Men   
3  1896  Athens  Aquatics   Swimming  MALOKINIS, Ioannis     GRE    Men   
4  1896  Athens  Aquatics   Swimming  CHASAPIS, Spiridon     GRE    Men   

                        Event   Medal  
0              100M Freestyle    Gold  
1              100M Freestyle  Silver  
2  100M Freestyle For Sailors  Bronze  
3  100M Freestyle For Sailors    Gold  
4  100M Freestyle For Sailors  Silver

In [2]:
sum(df[df['Medal'] == 'Bronze'].value_counts())

10369

Answer: 10369


Dalam olahraga air (Aquatics), kota mana yang memiliki rata-rata perolehan medali tertinggi?

In [3]:
a = df.groupby(['Sport','City'])['Medal'].value_counts()
a

Sport      City       Medal 
Aquatics   Amsterdam  Bronze    30
                      Gold      29
                      Silver    28
           Antwerp    Bronze    30
                      Silver    29
                                ..
Wrestling  Sydney     Gold      16
                      Silver    16
           Tokyo      Bronze    16
                      Gold      16
                      Silver    16
Name: Medal, Length: 1408, dtype: int64

Answer:

### **Soal 3 List Comprehension & Pandas** 

Pada Soal ini kalian **wajib** menggunakan feature list comprehension dan juga apply.<br>

In [39]:
import pandas as pd
df=pd.read_csv('Salaries.csv')
df

Id       EmployeeName  \
0            1     NATHANIEL FORD   
1            2       GARY JIMENEZ   
2            3     ALBERT PARDINI   
3            4  CHRISTOPHER CHONG   
4            5    PATRICK GARDNER   
...        ...                ...   
148649  148650      Roy I Tillery   
148650  148651       Not provided   
148651  148652       Not provided   
148652  148653       Not provided   
148653  148654          Joe Lopez   

                                              JobTitle    BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1                      CAPTAIN III (POLICE DEPARTMENT)  155966.02   
2                      CAPTAIN III (POLICE DEPARTMENT)  212739.13   
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC   77916.00   
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)  134401.60   
...                                                ...        ...   
148649                                       Custodian       0.00   
148650                                    Not provided        NaN   
148651                                    Not provided        NaN   
148652                                    Not provided        NaN   
148653                      Counselor, Log Cabin Ranch       0.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0              0.00  400184.25       NaN  567595.43         567595.43  2011   
1         245131.88  137811.38       NaN  538909.28         538909.28  2011   
2         106088.18   16452.60       NaN  335279.91         335279.91  2011   
3          56120.71  198306.90       NaN  332343.61         332343.61  2011   
4           9737.00  182234.59       NaN  326373.19         326373.19  2011   
...             ...        ...       ...        ...               ...   ...   
148649         0.00       0.00       0.0       0.00              0.00  2014   
148650          NaN        NaN       NaN       0.00              0.00  2014   
148651          NaN        NaN       NaN       0.00              0.00  2014   
148652          NaN        NaN       NaN       0.00              0.00  2014   
148653         0.00    -618.13       0.0    -618.13           -618.13  2014   

        Notes         Agency  Status  
0         NaN  San Francisco     NaN  
1         NaN  San Francisco     NaN  
2         NaN  San Francisco     NaN  
3         NaN  San Francisco     NaN  
4         NaN  San Francisco     NaN  
...       ...            ...     ...  
148649    NaN  San Francisco     NaN  
148650    NaN  San Francisco     NaN  
148651    NaN  San Francisco     NaN  
148652    NaN  San Francisco     NaN  
148653    NaN  San Francisco     NaN  

[148654 rows x 13 columns]

PT.A baru saja kena tegur dari negara karena ketauan tidak membayar uang lemburan. Anda sebagai seorang data Scientist diminta PT.A untuk mengolah data yang ada.

**Step pertama**

**Wajib Menggunakan .Apply()** <br>
PT.A melihat beberapa data tidak di isi oleh sebagian karyawan. kalian di minta untuk untuk merubah nilai 'Not provided' pada column `JobTitle` dan `EmployeeName` menjadi 'Error'.

In [58]:
df['JobTitle'] = df['JobTitle'].apply(lambda x: x[i]=='Error' if x[i] == 'Not Provided' else x[i] for i in x)

NameError: name 'x' is not defined

In [28]:
#jangan tulis code disini
#cell contoh hasil:

Id       EmployeeName  \
0            1     NATHANIEL FORD   
1            2       GARY JIMENEZ   
2            3     ALBERT PARDINI   
3            4  CHRISTOPHER CHONG   
4            5    PATRICK GARDNER   
...        ...                ...   
148649  148650      Roy I Tillery   
148650  148651              Error   
148651  148652              Error   
148652  148653              Error   
148653  148654          Joe Lopez   

                                              JobTitle    BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1                      CAPTAIN III (POLICE DEPARTMENT)  155966.02   
2                      CAPTAIN III (POLICE DEPARTMENT)  212739.13   
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC   77916.00   
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)  134401.60   
...                                                ...        ...   
148649                                       Custodian       0.00   
148650                                           Error        NaN   
148651                                           Error        NaN   
148652                                           Error        NaN   
148653                      Counselor, Log Cabin Ranch       0.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0              0.00  400184.25       NaN  567595.43         567595.43  2011   
1         245131.88  137811.38       NaN  538909.28         538909.28  2011   
2         106088.18   16452.60       NaN  335279.91         335279.91  2011   
3          56120.71  198306.90       NaN  332343.61         332343.61  2011   
4           9737.00  182234.59       NaN  326373.19         326373.19  2011   
...             ...        ...       ...        ...               ...   ...   
148649         0.00       0.00       0.0       0.00              0.00  2014   
148650          NaN        NaN       NaN       0.00              0.00  2014   
148651          NaN        NaN       NaN       0.00              0.00  2014   
148652          NaN        NaN       NaN       0.00              0.00  2014   
148653         0.00    -618.13       0.0    -618.13           -618.13  2014   

        Notes         Agency  Status  
0         NaN  San Francisco     NaN  
1         NaN  San Francisco     NaN  
2         NaN  San Francisco     NaN  
3         NaN  San Francisco     NaN  
4         NaN  San Francisco     NaN  
...       ...            ...     ...  
148649    NaN  San Francisco     NaN  
148650    NaN  San Francisco     NaN  
148651    NaN  San Francisco     NaN  
148652    NaN  San Francisco     NaN  
148653    NaN  San Francisco     NaN  

[148654 rows x 13 columns]

**Step Kedua**

**WAJIB Gunakan list comprehension**<br>
Anda diminta untuk menambahkan 200000 pada karyawan yang mendapat Overtime 0 pada Data yang telah dibersihkan di tahap 1.


In [ ]:
#your code goes here

In [29]:
#jangan tulis code disini
#cell contoh hasil:

Id       EmployeeName  \
0            1     NATHANIEL FORD   
1            2       GARY JIMENEZ   
2            3     ALBERT PARDINI   
3            4  CHRISTOPHER CHONG   
4            5    PATRICK GARDNER   
...        ...                ...   
148649  148650      Roy I Tillery   
148650  148651              Error   
148651  148652              Error   
148652  148653              Error   
148653  148654          Joe Lopez   

                                              JobTitle    BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1                      CAPTAIN III (POLICE DEPARTMENT)  155966.02   
2                      CAPTAIN III (POLICE DEPARTMENT)  212739.13   
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC   77916.00   
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)  134401.60   
...                                                ...        ...   
148649                                       Custodian       0.00   
148650                                           Error        NaN   
148651                                           Error        NaN   
148652                                           Error        NaN   
148653                      Counselor, Log Cabin Ranch       0.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         200000.00  400184.25       NaN  567595.43         567595.43  2011   
1         122565.94  137811.38       NaN  538909.28         538909.28  2011   
2         106088.18   16452.60       NaN  335279.91         335279.91  2011   
3          56120.71  198306.90       NaN  332343.61         332343.61  2011   
4           9737.00  182234.59       NaN  326373.19         326373.19  2011   
...             ...        ...       ...        ...               ...   ...   
148649    200000.00       0.00       0.0       0.00              0.00  2014   
148650          NaN        NaN       NaN       0.00              0.00  2014   
148651          NaN        NaN       NaN       0.00              0.00  2014   
148652          NaN        NaN       NaN       0.00              0.00  2014   
148653    200000.00    -618.13       0.0    -618.13           -618.13  2014   

        Notes         Agency  Status  
0         NaN  San Francisco     NaN  
1         NaN  San Francisco     NaN  
2         NaN  San Francisco     NaN  
3         NaN  San Francisco     NaN  
4         NaN  San Francisco     NaN  
...       ...            ...     ...  
148649    NaN  San Francisco     NaN  
148650    NaN  San Francisco     NaN  
148651    NaN  San Francisco     NaN  
148652    NaN  San Francisco     NaN  
148653    NaN  San Francisco     NaN  

[148654 rows x 13 columns]

**Tahap ketiga** 

**bebas memakai teknik apapun**<br>
selanjutnya ada kelasahan pada penulisan 'CAPTAIN III (POLICE DEPARTMENT)' ubahlah menjadi 'CAPTAIN III' saja

In [52]:
df['JobTitle'] = [df['JobTitle'] == df['JobTitle'][i] if df['JobTitle'][i] != 'CAPTAIN III (POLICE DEPARTMENT)' else df['JobTitle'][i] == 'CAPTAIN III' for i in df['JobTitle']]

KeyError: 'GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY'

In [30]:
#jangan tulis code disini
#cell contoh hasil:

Id       EmployeeName  \
0            1     NATHANIEL FORD   
1            2       GARY JIMENEZ   
2            3     ALBERT PARDINI   
3            4  CHRISTOPHER CHONG   
4            5    PATRICK GARDNER   
...        ...                ...   
148649  148650      Roy I Tillery   
148650  148651              Error   
148651  148652              Error   
148652  148653              Error   
148653  148654          Joe Lopez   

                                              JobTitle    BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1                                          CAPTAIN III  155966.02   
2                                          CAPTAIN III  212739.13   
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC   77916.00   
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)  134401.60   
...                                                ...        ...   
148649                                       Custodian       0.00   
148650                                           Error        NaN   
148651                                           Error        NaN   
148652                                           Error        NaN   
148653                      Counselor, Log Cabin Ranch       0.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         200000.00  400184.25       NaN  567595.43         567595.43  2011   
1         122565.94  137811.38       NaN  538909.28         538909.28  2011   
2         106088.18   16452.60       NaN  335279.91         335279.91  2011   
3          56120.71  198306.90       NaN  332343.61         332343.61  2011   
4           9737.00  182234.59       NaN  326373.19         326373.19  2011   
...             ...        ...       ...        ...               ...   ...   
148649    200000.00       0.00       0.0       0.00              0.00  2014   
148650          NaN        NaN       NaN       0.00              0.00  2014   
148651          NaN        NaN       NaN       0.00              0.00  2014   
148652          NaN        NaN       NaN       0.00              0.00  2014   
148653    200000.00    -618.13       0.0    -618.13           -618.13  2014   

        Notes         Agency  Status  
0         NaN  San Francisco     NaN  
1         NaN  San Francisco     NaN  
2         NaN  San Francisco     NaN  
3         NaN  San Francisco     NaN  
4         NaN  San Francisco     NaN  
...       ...            ...     ...  
148649    NaN  San Francisco     NaN  
148650    NaN  San Francisco     NaN  
148651    NaN  San Francisco     NaN  
148652    NaN  San Francisco     NaN  
148653    NaN  San Francisco     NaN  

[148654 rows x 13 columns]

### **Soal 4 One Hot Encoding**

Buatlah **sebuah fungsi** bernama `one_hot`, yang menerima input sebuah numpy array.

Ex:<br>
input array [2,3,1,2] maka hasil dari fungsi kalian:

array [[0,1,0], [0,0,1], [1,0,0], [0,1,0]]

Constrain: <br>

- Integers dalam input selalu di awali 1. 
- Integers dalam input hanya positif dan bukan 0.
- Integers dalam input tidak akan skip angka.

In [53]:
import numpy as np
input_array = np.array([3,2,1,2,3,3,3])
#your code goes here
import numpy as np

def one_hot(x):
    encoded_array = np.zeros((x.size, x.max()+1), dtype=int)
    encoded_array[np.arange(x.size),x] = 1
    
    return encoded_array

In [54]:
one_hot(input_array)

array([[0, 0, 0, 1],
       [0, 0, 1, 0],
       [0, 1, 0, 0],
       [0, 0, 1, 0],
       [0, 0, 0, 1],
       [0, 0, 0, 1],
       [0, 0, 0, 1]])

In [3]:
#jangan tulis code disini
#cell contoh hasil:

array([[0., 0., 1.],
       [0., 1., 0.],
       [1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.],
       [0., 0., 1.],
       [0., 0., 1.]])